woocommerce/woocommerce-ios
# [Mobile Payments] Add Universal Link support for Set up Tap to Pay (and Collect Payment)

Part of: #9370


## Description


In order to promote Tap to Pay and get more users to try it, we will market it using JITMs, emails, and other direct communication. It would be helpful in that communication to take merchants directly to the `Set up Tap to Pay on iPhone` flow.

This PR adds support for that universal link (woocommerce.com/mobile/payments/tap-to-pay) and also a collect payment link. 

### Changes to Universal Link handling
I made some changes to the API of the Universal Link Router, so that a `Route` can handle more than one subpath. This allows us to route all the payments related links together, via the `MainTabBarController` and the `HubMenuCoordinator`.

Instead of matching a single subpath, each `Route` has a `canHandle` method which the `RouteMatcher` calls to determine which Route to use (still the first which returns `true`.) 

The `performAction` method has the subpath as an additional parameter, so that it can route the action appropriately.

Passing the path substrings around would likely lead to issues, so once we are performing the action we handle specific `DeepLinkDestination`s which are defined in the class that handles those destinations.

## Testing instructions


1. Run the app
2. Go to a system app, such as Notes or Reminders, and add an entry with each universal link that we support as URLs
3. Tap on the URL
4. Observe that you're taken to the appropriate screen in the app.

URLs:
- Payments – https://woocommerce.com/mobile/payments
- Set up Tap to Pay on iPhone – https://woocommerce.com/mobile/payments/tap-to-pay
- Collect Payment – https://woocommerce.com/mobile/payments/collect-payment
- Order Details – https://woocommerce.com/mobile/orders/details?blog_id={yourSiteID}&order_id={anOrderIDFromYourSite}

Observe that these URLs work with, and without, a trailing `/`

Note that the link to Set up Tap to Pay on iPhone assumes that you're passing onboarding, but does not handle failures of onboarding, which is temporary pending work to show onboarding in the flow when it's required. This behaviour is feature-flagged, so the link will not be released until the onboarding work is done.

## Screenshots



https://user-images.githubusercontent.com/2472348/230609051-acb80ec8-6662-4de0-a337-033208456eb2.mp4


---
